contributors/design-proposals/auth/kubectl-exec-plugins.md

```diff
@@ -0,0 +1,264 @@
+      command: /bin/kubectl-login
+      args: ["hello", "world"]
+```
+
+The AWS authenticator, modified to return structured output, would become:
+
+```
+users:
+- name: kubernetes-admin
+  user:
+    exec:
+      apiVersion: "client.authentication.k8s.io/v1alpha1"
+      command: heptio-authenticator-aws
+      # CLUSTER_ID and ROLE_ARN should be replaced with actual desired values.
+      args: ["token", "-i", "(CLUSTER_ID)", "-r", "(ROLE_ARN)"]
+```
+
+## TLS client certificate support
+
+TLS client certificate support is orthogonal to bearer tokens, but something that
+we should consider supporting in the future. Beyond requiring different command
+output, it also requires changes to the client-go `AuthProvider` interface.
+
+The current The auth provider interface doesn't let the user modify the dialer,
+only wrap the transport.
+
+```
+type AuthProvider interface {
+    // WrapTransport allows the plugin to create a modified RoundTripper that
+    // attaches authorization headers (or other info) to requests.
+    WrapTransport(http.RoundTripper) http.RoundTripper
+    // Login allows the plugin to initialize its configuration. It must not
+    // require direct user interaction.
+    Login() error
+}
+```
+
+Since this doesn't let a `AuthProvider` supply things like client certificates,
+the signature of the `AuthProvider` should change too ([with corresponding changes
+to `k8s.io/client-go-transport`](
+https://gist.github.com/ericchiang/7f5804403b359ebdf79dcf76c4071bff)):
+
+```go
+import (
+    "k8s.io/client-go/transport"
+    // ...
+)
+
+type AuthProvider interface {
+    // UpdateTransportConfig updates a config by adding a transport wrapper,
+    // setting a bearer token (should ignore if one is already set), or adding
+    // TLS client certificate credentials.
+    //
+    // This is called once on transport initialization. Providers that need to
+    // rotate credentials should use Config.WrapTransport to dynamically update
+    // credentials.
+    UpdateTransportConfig(c *transport.Config)
+
+    // Login() dropped, it was never used.
+}
+```
+
+This would let auth transports supply TLS credentials, as well as instrument
+transports with in-memory rotation code like the utilities implemented by
+[`k8s.io/client-go/util/certificate`](https://godoc.org/k8s.io/client-go/util/certificate).
+
+The `ExecCredentials` would then expand to provide TLS options.
+
+```go
+type ExecCredentials struct {
+    metav1.TypeMeta `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+    // Token is a bearer token used by the client for request authentication.
+    Token string `json:"token,omitempty"`
+    // Expiry indicates a unix time when the provided credentials expire.
+    Expiry int64 `json:"expiry,omitempty"`
+
+    ClientCertificateData []byte `json:"clientCertificateData,omitempty"`
+    ClientKeyData         []byte `json:"clientKeyData,omitempty"`
```

Comment:
if we spec the cert/key data must be in PEM format, it can be strings, which makes it easier for something which already has PEM data to pass it on without needing a utility like `base64`